esi_leap/manager/service.py

```diff
@@ -75,8 +77,14 @@ def _expire_contracts(self):
         now = timeutils.utcnow()
         for c in contracts:
             if c.end_time <= now:
-                LOG.info("Expiring contract %s", c.uuid)
-                c.expire(self._context)
+
+                with lockutils.lock(c.uuid,
+                                    lock_file_prefix='contract',
+                                    external=True):
+                    c_locked = contract.Contract.get_by_uuid(c.uuid)
+                    if c_locked.status != statuses.EXPIRED:
+                        c_locked.expire(self._context)
+                        LOG.info("Expiring contract %s", c.uuid)
```

Comment:
Given that there are similar locks here, and in the offer expire method - would it make sense to have the lock be in the actual contract expire method?

Reply:
You cannot do this for the reason I stated below.